diff --git a/docroot/.htaccess b/docroot/.htaccess
index 6a1c1a38e..d5b602e3a 100644
--- a/docroot/.htaccess
+++ b/docroot/.htaccess
@@ -162,6 +162,19 @@ AddEncoding gzip svgz
   RewriteCond %{HTTP_HOST} iihr\.((prod|stage|dev)\.drupal\.)?uiowa\.edu$ [NC]
   RewriteRule ^igs/publications/(.*)$ https://igs.iihr.uiowa.edu/igs/publications/$1 [R=301,L]
 
+  # Redirect rules for medicine
+  RewriteCond %{HTTP_HOST} ^(www\.)?medicine\.((prod|stage|dev)\.drupal\.)?uiowa\.edu$ [NC]
+  RewriteRule ^iowaprotocols/?$ https://iowaprotocols.medicine.uiowa.edu [R=301,L]
+
+  RewriteCond %{HTTP_HOST} ^(www\.)?medicine\.((prod|stage|dev)\.drupal\.)?uiowa\.edu$ [NC]
+  RewriteRule ^iowaprotocols/content(/.*)?$ https://iowaprotocols.medicine.uiowa.edu/protocols$1 [R=301,L]
+
+  RewriteCond %{HTTP_HOST} ^(www\.)?medicine\.((prod|stage|dev)\.drupal\.)?uiowa\.edu$ [NC]
+  RewriteRule ^iowaprotocols/(.+)$ https://iowaprotocols.medicine.uiowa.edu/protocols/$1 [R=301,L]
+
+
+
+
   # Set "protossl" to "s" if we were accessed via https://.  This is used later
   # if you enable "www." stripping or enforcement, in order to ensure that
   # you don't bounce between http and https.
