diff --git a/docroot/.htaccess b/docroot/.htaccess
index bca408037..1a07faabe 100644
--- a/docroot/.htaccess
+++ b/docroot/.htaccess
@@ -49,8 +49,8 @@ AddEncoding gzip svgz
   # Enable expirations.
   ExpiresActive On
 
-  # Cache all files for 1 year after access.
-  ExpiresDefault "access plus 1 year"
+  # Cache all files for 10 minutes after access.
+  ExpiresDefault "access plus 10 minutes"
 
   <FilesMatch \.php$>
     # Do not allow PHP scripts to be cached unless they explicitly send cache
