diff --git a/core/misc/announce.js b/core/misc/announce.js
index 32368b2a0c0a58f788b3e74367d7d06e752c85e0..d983070273cf752615dc7e295e0601a7f6cc7f45 100644
--- a/core/misc/announce.js
+++ b/core/misc/announce.js
@@ -39,6 +39,7 @@
         liveElement.className = 'visually-hidden';
         liveElement.setAttribute('aria-live', 'polite');
         liveElement.setAttribute('aria-busy', 'false');
+        liveElement.setAttribute('aria-atomic', 'false');
         document.body.appendChild(liveElement);
       }
     },
@@ -50,6 +51,7 @@
   function announce() {
     const text = [];
     let priority = 'polite';
+    let atomic = 'false';
     let announcement;

     // Create an array of announcement strings to be joined and appended to the
@@ -59,9 +61,10 @@
       announcement = announcements.pop();
       text.unshift(announcement.text);
       // If any of the announcements has a priority of assertive then the group
-      // of joined announcements will have this priority.
+      // of joined announcements will have this priority. Also add aria-atomic='true' for full Web Content Accessibility Guidelines compliance
       if (announcement.priority === 'assertive') {
         priority = 'assertive';
+        atomic = 'true';
       }
     }

@@ -72,6 +75,8 @@
       liveElement.setAttribute('aria-busy', 'true');
       // Set the priority to assertive, or default to polite.
       liveElement.setAttribute('aria-live', priority);
+      // Set the aria-atomic value to true, or default to false
+      liveElement.setAttribute('aria-atomic', atomic);
       // Print the text to the live region. Text should be run through
       // Drupal.t() before being passed to Drupal.announce().
       liveElement.innerHTML = text.join('\n');
@@ -96,7 +101,6 @@
    * @param {string} [priority='polite']
    *   A string to indicate the priority of the message. Can be either
    *   'polite' or 'assertive'.
-   *
    * @return {function}
    *   The return of the call to debounce.
    *
