diff --git a/core/lib/Drupal/Core/Form/FormBuilder.php b/core/lib/Drupal/Core/Form/FormBuilder.php
index cec9816..e35c0e9 100644
--- a/core/lib/Drupal/Core/Form/FormBuilder.php
+++ b/core/lib/Drupal/Core/Form/FormBuilder.php
@@ -1254,7 +1254,7 @@ protected function handleInputElement($form_id, &$element, FormStateInterface &$
         if (!$input_exists && !$form_state->isRebuilding() && !$form_state->isProgrammed()) {
           // Add the necessary parent keys to FormState::$input and sets the
           // element's input value to NULL.
-          NestedArray::setValue($form_state->getUserInput(), $element['#parents'], NULL);
+          NestedArray::setValue($form_state->getUserInput(), $element['#parents'], NULL, TRUE);
           $input_exists = TRUE;
         }
         // If we have input for the current element, assign it to the #value
