diff --git a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
index 5234c88d90..acdd8b19a9 100644
--- a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
+++ b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
@@ -478,7 +478,7 @@ class MediaLibraryWidget extends WidgetBase implements ContainerFactoryPluginInt
 
     // Add a button that will load the Media library in a modal using AJAX.
     $element['open_button'] = [
-      '#type' => 'submit',
+      '#type' => 'button',
       '#value' => $this->t('Add media'),
       '#name' => $field_name . '-media-library-open-button' . $id_suffix,
       '#attributes' => [
